src/client/Client.cc

```diff
@@ -7496,73 +7501,169 @@ int Client::_do_setattr(Inode *in, struct ceph_statx *stx, int mask,
-      /* Must squash the any setuid/setgid bits with an ownership change */
-      in->mode &= ~(S_ISUID|S_ISGID);
-      in->mark_caps_dirty(CEPH_CAP_AUTH_EXCL);
+    } else if (!in->caps_issued_mask(CEPH_CAP_AUTH_SHARED) ||
+               in->mode != stx->stx_mode) {
+      args.setattr.mode = stx->stx_mode;
+      inode_drop |= CEPH_CAP_AUTH_SHARED;
     }
+  } else if (in->caps_issued_mask(CEPH_CAP_AUTH_EXCL) &&
+             kill_sguid && S_ISREG(in->mode) &&
+	     (in->mode & (S_IXUSR|S_IXGRP|S_IXOTH))) {
+    /* Must squash the any setuid/setgid bits with an ownership change */
+    in->mode &= ~(S_ISUID|S_ISGID);
+    in->mark_caps_dirty(CEPH_CAP_AUTH_EXCL);
+  }
 
-    if (mask & CEPH_SETATTR_BTIME) {
+  if (mask & CEPH_SETATTR_BTIME) {
+    ldout(cct,10) << "changing btime to " << in->btime << dendl;
+
+    if (in->caps_issued_mask(CEPH_CAP_AUTH_EXCL)) {
       in->ctime = ceph_clock_now();
       in->cap_dirtier_uid = perms.uid();
       in->cap_dirtier_gid = perms.gid();
       in->btime = utime_t(stx->stx_btime);
       in->mark_caps_dirty(CEPH_CAP_AUTH_EXCL);
       mask &= ~CEPH_SETATTR_BTIME;
-      ldout(cct,10) << "changing btime to " << in->btime << dendl;
+    } else if (!in->caps_issued_mask(CEPH_CAP_AUTH_SHARED) ||
+               in->btime != utime_t(stx->stx_btime)) {
+      args.setattr.btime = utime_t(stx->stx_btime);
+      inode_drop |= CEPH_CAP_AUTH_SHARED;
+    }
+  }
+
+  if (mask & CEPH_SETATTR_SIZE) {
+    if ((uint64_t)stx->stx_size >= mdsmap->get_max_filesize()) {
+      //too big!
+      ldout(cct,10) << "unable to set size to " << stx->stx_size << ". Too large!" << dendl;
+      return -CEPHFS_EFBIG;
+    }
+
+    ldout(cct,10) << "changing size to " << stx->stx_size << dendl;
+    if (in->caps_issued_mask(CEPH_CAP_FILE_EXCL) &&
+        !(mask & CEPH_SETATTR_KILL_SGUID) &&
+        stx->stx_size >= in->size) {
+      if (stx->stx_size > in->size) {
+        in->size = in->reported_size = stx->stx_size;
+        in->cap_dirtier_uid = perms.uid();
+        in->cap_dirtier_gid = perms.gid();
+        in->mark_caps_dirty(CEPH_CAP_FILE_EXCL);
+        mask &= ~(CEPH_SETATTR_SIZE);
+        mask |= CEPH_SETATTR_MTIME;
+      } else {
+        // ignore it when size doesn't change
+      }
+    } else {
+      args.setattr.size = stx->stx_size;
+      inode_drop |= CEPH_CAP_FILE_SHARED | CEPH_CAP_FILE_RD |
+                    CEPH_CAP_FILE_WR;
+    }
+  }
+
+  if (mask & CEPH_SETATTR_MTIME) {
+    if (in->caps_issued_mask(CEPH_CAP_FILE_EXCL)) {
+      in->mtime = utime_t(stx->stx_mtime);
+      in->ctime = ceph_clock_now();
+      in->cap_dirtier_uid = perms.uid();
+      in->cap_dirtier_gid = perms.gid();
+      in->time_warp_seq++;
+      in->mark_caps_dirty(CEPH_CAP_FILE_EXCL);
+      mask &= ~CEPH_SETATTR_MTIME;
+    } else if (in->caps_issued_mask(CEPH_CAP_FILE_WR) &&
+               utime_t(stx->stx_mtime) > in->mtime) {
```

Comment:
This one concerns me. Would this not break a reader + writer where the reader is polling for mtime updates?

Reply:
Yeah, this sounds a problem. Anyway in this case we must update the `mtime` to the MDS by either setattr request or cap update request.

To be simple I will remove this code currently. Thanks.

Reply:
There is some general confusion around what caps cover these attributes. The kernel wants Fs to read the mtime in a getattr request:
```
        if (want & (STATX_ATIME|STATX_MTIME|STATX_CTIME|STATX_SIZE|
                    STATX_BLOCKS))
                mask |= CEPH_CAP_FILE_SHARED;
```
...but that looks wrong, in hindsight. Can a client hold Fw caps while another has Fs? I suspect that it can. If so, then the kclient's getattr path is not working as it should.

Again -- we *really* need a canonical treatise on what caps cover what inode metadata (particularly for FILE caps). Maybe we should badger @liewegas, @gregsfortytwo and @ukernel for such a thing?

My guess though is that this has never been clearly defined throughout cephfs's existence. We'll probably need to write up a "spec" for this ourselves, and then fix the userland and kernel code to conform to it.

Reply:
@batrick @vshankar Please take a look.

As Jeff mentioned and I double checked, the `Fs` and `Fw` caps are not allowed to distinct clients at the same time.

If one reader is polling the `mtime` updates it should use the `statx()` with `AT_STATX_FORCE_SYNC`. From the posix doc, without the `AT_STATX_FORCE_SYNC`, it won't make sure that the `mtime` will be the newest,  just return what the local cache has.

Once the `AT_STATX_FORCE_SYNC` was set, client will always sync the attributes from the MDS server by requiring the `Fs` caps. Then in MDS it will rdlock the FILE lock, that means all the `Fw` caps must be revoked. Then the writer client will flush it.